lib/utilization/docker-info.js

```diff
@@ -5,61 +5,162 @@
+    if (err == null) {
+      // The boot id proc file exists, so use it to get the container id.
+      return common.readProc(BOOT_ID_PROC_FILE, (_, data, cbAgent = agent) => {
+        readProcBootId({ data, agent: cbAgent, callback })
+      })
     }
 
-    data = data.trim()
-    const asciiData = Buffer.from(data, 'ascii').toString()
+    logger.debug('Container boot id is not available in cgroups info')
 
-    if (data !== asciiData) {
-      bootIdError()
+    if (hasAwsContainerApi() === false) {
+      // We don't seem to have a recognized location for getting the container
+      // identifier.
+      logger.debug('Container is not in a recognized ECS container, omitting boot info')
+      recordBootIdError(agent)
       return callback(null, null)
     }
 
-    if (data.length !== 36) {
-      bootIdError()
-      if (data.length > 128) {
-        data = data.substring(0, 128)
+    getEcsContainerId({ agent, callback, logger })
+  })
+}
+
+/**
+ * Queries the AWS ECS metadata API to get the boot id.
+ *
+ * @param {object} params Function parameters.
+ * @param {object} params.agent Newrelic agent instance.
+ * @param {Function} params.callback Typical error first callback. The second
+ * parameter is the boot id as a string.
+ * @param {object} [params.logger] Internal logger instance.
+ */
+function getEcsContainerId({ agent, callback, logger }) {
+  const ecsApiUrl =
+    process.env.ECS_CONTAINER_METADATA_URI_V4 || process.env.ECS_CONTAINER_METADATA_URI
+  const req = http.request(ecsApiUrl, (res) => {
+    let body = Buffer.alloc(0)
+    res.on('data', (chunk) => {
+      body = Buffer.concat([body, chunk])
+    })
+    res.on('end', () => {
+      try {
+        const json = body.toString('utf8')
+        const data = JSON.parse(json)
+        if (data.DockerId == null) {
+          logger.debug('Failed to find DockerId in response, omitting boot info')
+          recordBootIdError(agent)
+          return callback(null, null)
+        }
+        callback(null, data.DockerId)
+      } catch (error) {
+        logger.debug('Failed to process ECS API response, omitting boot info: ' + error.message)
+        recordBootIdError(agent)
+        callback(null, null)
       }
-    }
+    })
+  })
 
-    return callback(null, data)
+  req.on('error', () => {
+    logger.debug('Failed to query ECS endpoint, omitting boot info')
+    recordBootIdError(agent)
+    callback(null, null)
   })
 
-  function bootIdError() {
-    agent.metrics.getOrCreateMetric(NAMES.UTILIZATION.BOOT_ID_ERROR).incrementCallCount()
+  req.end()
+}
+
+/**
+ * Inspects the running environment to determine if the AWS ECS metadata API
+ * is available.
+ *
+ * @see https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ec2-metadata.html
+ *
+ * @returns {boolean}
+ */
+function hasAwsContainerApi() {
+  if (process.env.ECS_CONTAINER_METADATA_URI_V4 != null) {
+    return true
   }
+  return process.env.ECS_CONTAINER_METADATA_URI != null
```

Comment:
```suggestion
  return process.env.ECS_CONTAINER_METADATA_URI_V4 != null || process.env.ECS_CONTAINER_METADATA_URI != null
```

Reply:
For me, it's clearer as it is written that we are preferring the `V4` version and then falling back to the other.